[LoopUnroll] Ignore inlinable calls if unrolling is forced

gburgessiv opened the pull request:
`NumInlineCandidates` counts candidates that are _very likely_ to be inlined. This is a useful metric, but causes linker warnings if:
- the loop to be unrolled has had unrolling forced by the user, and
- the inliner fails to inline the call (e.g., because it's considered a very cold callsite)

Fixes #88141

gburgessiv commented:
Hey Nikita, I see you've done a few reviews in this area. Would you kindly be able to take a look at this CL?

fhahn commented:
> This is a useful metric, but causes linker warnings if:

From the description, it's not clear what kind of warnings this would yield. Looking at the linked issue, it's remarks that get printed by default (`loop not unrolled: the optimizer was unable to perform the requested transformation; the transformation might be disabled or specified as part of an unsupported transformation ordering`). 

Would be good to clarify in the description.

Regardless of the patch at hand, I don't think we really should print those messages by default; there are other reasons why the compiler won't unroll a loop even if forced by the user.

Could you also add a link to https://github.com/llvm/llvm-project/issues/88141 in the description, not in the title as ` (#88141) `? This could be confusing as Github will add the PR ID in a similar fashion when landing.

nikic commented on llvm/test/Transforms/LoopUnroll/unroll-calls-if-forced.ll:
```diff
@@ -0,0 +1,35 @@
+; RUN: opt -passes=loop-unroll -disable-verify -S < %s | FileCheck %s
```
Why does this `-disable-verify`?

Please also use UTC.

gburgessiv replied:
> Why does this -disable-verify?

Copy-paste from another test; I removed this.

> Please also use UTC.

I assume for commit timestamps? Done.

aeubanks replied:
UTC as in llvm/utils/update_test_checks.py to automatically generate CHECK lines

gburgessiv replied:
Ah, that makes more sense. Thanks

nikic commented:
I think the change makes sense conceptually -- pragma unroll should override pretty much all inlining heuristics, and I don't see a strong reason why it wouldn't override this one as well.

A possible caveat is the interaction with the pre-link pipeline: Are we okay with unrolling the calls there, or would it be better to wait until post-link (in which case the call has likely been inlined)?

gburgessiv committed:
[LoopUnroll] Ignore inlinable calls if unrolling is forced

`NumInlineCandidates` counts candidates that are _very likely_ to be
inlined. This is a useful metric, but causes linker warnings if:
- the loop to be unrolled has had unrolling forced by the user, and
- the inliner fails to inline the call (e.g., because it's considered a
  very cold callsite)

Fixes #88141

aeubanks commented:
reverse ping, Chromium is seeing something similar that may be fixed by this

gburgessiv commented:
Ah, sorry for the delay here.

> A possible caveat is the interaction with the pre-link pipeline: Are we okay with unrolling the calls there, or would it be better to wait until post-link (in which case the call has likely been inlined)?

I added a commit that tries to funnel `PrepareForLTO` through properly. I'm not confident it caught everything, but it should be a step toward addressing this, at least. I agree pushing this out of prelink is good, since the prelink pass manager seems to favor transformations that reduce IR size.

aeubanks commented on llvm/lib/Passes/PassBuilderPipelines.cpp:
```diff
@@ -481,9 +481,10 @@ PassBuilder::buildO1FunctionSimplificationPipeline(OptimizationLevel Level,
+    LPM2.addPass(LoopFullUnrollPass(
+        Level.getSpeedupLevel(),
+        /* OnlyWhenForced= */ !PTO.LoopUnrolling, PTO.ForgetAllSCEVInLoopUnroll,
+        /* PrepareForLTO= */ isLTOPreLink(ThinOrFullLTOPhase::ThinLTOPreLink)));
```
`isLTOPreLink(Phase)`, ditto below

aeubanks reviewed (commented):
the PrepareForLTO portion needs testing. you'll need to add parsing code for something like `-passes='loop-unroll-full<prepare-for-lto>'`, grep for `parseLoopRotateOptions` for an example

also, isn't this still going to result in the remark getting printed in the pre-link optimization pipeline? I think we should treat the PrepareForLTO and the call heuristic as two separate issues and go back to the original PR, splitting out the PrepareForLTO change

please move the issue being fixed from the commit title to the commit message like `Fixes #88141`, otherwise it's confusing with the PR number

gburgessiv commented:
> also, isn't this still going to result in the remark getting printed in the pre-link optimization pipeline?

I only observed the remarks during LTO, so I'd assumed that the prelink pipeline has a way of bypassing these checks (which passes my smell test, at least, since it's reasonable to expect opts to be in an unfinished state after pre-link).

> I think we should treat the PrepareForLTO and the call heuristic as two separate issues and go back to the original PR, splitting out the PrepareForLTO change

Yeah, fixing the bug first SGTM.

> please move the issue being fixed from the commit title to the commit message like Fixes #88141, otherwise it's confusing with the PR number

Done

LeszekSwirski commented:
Is there any progress here? We've had to disable this annotation in V8 for now because of this issue (crbug.com/377144577)

gburgessiv commented:
Friendly ping on this reduced PR; seems V8 still has a workaround locally for this issue.
